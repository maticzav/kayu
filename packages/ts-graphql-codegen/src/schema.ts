import * as fs from 'fs'
import {
  getIntrospectionQuery,
  IntrospectionQuery,
  IntrospectionSchema,
} from 'graphql'
import fetch from 'node-fetch'
import * as path from 'path'
import { promisify } from 'util'

const readfile = promisify(fs.readFile)

export type LoadSchemaFromOptions = {
  /**
   * Dictionary of headers where key represents header name
   * and value represents header value.
   */
  headers?: { [key: string]: string }
  /**
   * The method used to perform the request.
   */
  method?: 'POST'
}

/**
 * Loads introspection schema that we use to generate the library
 * from a specified endpoint.
 *
 * We assume that return value is correct and don't check it
 * beyond making it conform to type.
 */
export async function loadSchemaFromURL(
  endpoint: string,
  opts?: LoadSchemaFromOptions,
): Promise<IntrospectionSchema> {
  /* Construct request. */
  const body = {
    query: getIntrospectionQuery(),
  }

  /* Fetch the response */
  const res = await fetch(endpoint, {
    body: JSON.stringify(body),
    headers: {
      ...opts?.headers,
      'Content-Type': 'application/json',
    },
    method: opts?.method || 'POST',
  })

  const query: { data: IntrospectionQuery } = await res.json()
  return query.data.__schema
}

/**
 * Loads the schema from local file path. Path should be given as a
 * global - you may use `path` module to get it.
 *
 * We assume that file content is correct and perform no deep checks
 * on type signature of the file.
 */
export async function loadSchemaFromPath(
  filepath: string,
): Promise<IntrospectionSchema> {
  /* Check that path is absolute. */
  if (!path.isAbsolute(filepath)) {
    throw new Error(`Expected absolute path, but got ${filepath}.`)
  }

  /* Laod the data. */
  const data = await readfile(filepath)
  const schema: IntrospectionSchema = JSON.parse(data.toString())

  return schema
}
