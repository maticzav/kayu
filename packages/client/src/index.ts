export { Codec } from './codec'

/**
 * These exports are meant to be used solely by the generated
 * code. Do not use them anywhere. These functions are undocumented
 * and might change without breaking change notice.
 *
 * Instead, use functions that have been documented.
 */
export { OperationType } from './document'
export { Argument, arg } from './document/argument'
export { Field, composite, fragment, leaf } from './document/field'
export { nullable, list } from './internal/decode'
export { Fields, SelectionSet, selection } from './selection'
export { perform, PerformInput, PerformOutput } from './http'
