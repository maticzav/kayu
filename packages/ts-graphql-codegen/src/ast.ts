import {
  IntrospectionInputType,
  IntrospectionInputTypeRef,
  IntrospectionListTypeRef,
  IntrospectionNamedTypeRef,
  IntrospectionNonNullTypeRef,
  IntrospectionOutputType,
  IntrospectionOutputTypeRef,
  IntrospectionType,
  IntrospectionTypeRef,
} from 'graphql'

/**
 * Unwrapps the nested named type from the refernce.
 */
export function getNamedTypeRef<T>(
  ref: IntrospectionTypeRef,
): IntrospectionNamedTypeRef {
  switch (ref.kind) {
    case 'LIST': {
      return getNamedTypeRef(ref.ofType)
    }
    case 'NON_NULL': {
      return getNamedTypeRef(ref.ofType)
    }
    case 'INPUT_OBJECT':
    case 'ENUM':
    case 'UNION':
    case 'OBJECT':
    case 'SCALAR':
    case 'INTERFACE': {
      return ref
    }
  }
}

/**
 * Inverted references help us calculate data some types more easily
 * since typescript, by default, has non-null fields while GraphQL is
 * nullable by default.
 */

export interface IntrospectionNullableTypeRef<
  T extends IntrospectionTypeRef = IntrospectionTypeRef
> {
  readonly kind: 'NULLABLE'
  readonly ofType: T
}

export type IntrospectionInvertedTypeRef<T extends IntrospectionType> =
  | IntrospectionNamedTypeRef<T>
  | IntrospectionInvertedListTypeRef<any>
  | IntrospectionNullableTypeRef<
      IntrospectionNamedTypeRef<T> | IntrospectionListTypeRef<any>
    >

export interface IntrospectionInvertedListTypeRef<
  T extends IntrospectionInvertedTypeRef = IntrospectionInvertedTypeRef<any>
> {
  readonly kind: 'LIST'
  readonly ofType: T
}

export type IntrospectionInvertedOutputTypeRef = IntrospectionInvertedTypeRef<IntrospectionOutputType>

export type IntrospectionInvertedInputTypeRef = IntrospectionInvertedTypeRef<IntrospectionInputType>

/**
 * Inverts the ref to inverted instance to make type calculations easier.
 */
export function invert<T>(
  ref: IntrospectionOutputTypeRef,
): IntrospectionInvertedOutputTypeRef {
  /**
   * Every field is by default nullable (read GraphQL spec).
   * If it's wrapped to be non-nullable we simply remove the
   * nullablility wrapper that we added down in the chain.
   */
  switch (ref.kind) {
    /* Non Nullable */
    case 'NON_NULL': {
      const inverted = invert(ref.ofType)
      switch (inverted.kind) {
        case 'NULLABLE': {
          return inverted.ofType
        }
        default: {
          return inverted
        }
      }
    }
    /* Named */
    case 'LIST':
    case 'ENUM':
    case 'UNION':
    case 'OBJECT':
    case 'SCALAR':
    case 'INTERFACE': {
      return {
        kind: 'NULLABLE',
        ofType: ref,
      }
    }
  }
}

/**
 * Inverts the ref back to its original state.
 */
export function revert<T>(
  ref: IntrospectionInvertedOutputTypeRef,
): IntrospectionOutputTypeRef {
  /**
   * Every reference is now by default non-nullalbe,
   * that's why we should make every instance non-nullable
   * and remove the wrapper if we come across nullable wrapper.
   *
   * Note that we are simply changing perspective, not the values themself.
   */
  switch (ref.kind) {
    /* Nullable */
    case 'NULLABLE': {
      const reverted = revert(ref.ofType)
      switch (reverted.kind) {
        case 'NON_NULL': {
          return reverted.ofType
        }
        default: {
          return reverted
        }
      }
    }
    /* Named */
    case 'LIST':
    case 'ENUM':
    case 'UNION':
    case 'OBJECT':
    case 'SCALAR':
    case 'INTERFACE': {
      return {
        kind: 'NON_NULL',
        ofType: ref,
      }
    }
  }
}
