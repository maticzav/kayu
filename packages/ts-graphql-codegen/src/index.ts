export { GQLGenerator } from './generator'
export {
  loadSchemaFromPath,
  LoadSchemaFromOptions,
  loadSchemaFromURL,
} from './schema'
