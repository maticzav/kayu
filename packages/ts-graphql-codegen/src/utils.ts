import * as camelcase from 'camelcase'

/**
 * Utility function that we use to make sure switch statements
 * are exhaustive.
 */
export function never(_x: never): never {
  throw new Error('Switch statement not exhaustive!')
}

/**
 * Checks that the given value is not undefined.
 */
export function defined<T>(val: T | undefined): val is T {
  return val !== undefined
}

/**
 * Returns the camelCased string.
 */
export function camel(val: string): string {
  return camelcase(val)
}

/**
 * Returns the PascalCased string.
 */
export function pascal(val: string): string {
  return camelcase(val, { pascalCase: true })
}
