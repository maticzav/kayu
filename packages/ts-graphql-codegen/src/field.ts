/**
 * This file contains helper methods that we use for code generation
 * and serve a single purpose - that's why I extracted them here.
 */
import { IntrospectionTypeRef } from 'graphql'

import { IntrospectionInvertedTypeRef, invert } from './ast'

/**
 * Wraps the given type using a given reference.
 */
export function wrap(type: string, ref: IntrospectionTypeRef): string {
  return wrapInverted(type, invert(ref))
}

/**
 * This is an internal helper function that lets us more easily define the return type
 * of the refered one.
 */
function wrapInverted(type: string, ref: IntrospectionInvertedTypeRef): string {
  switch (ref.kind) {
    /* Nullable */
    case 'NULLABLE': {
      return `(${wrapInverted(type, ref.ofType)} | null)`
    }
    /* List */
    case 'LIST': {
      return `(${wrapInverted(type, ref.ofType)}[])`
    }
    /* Named */
    case 'ENUM':
    case 'UNION':
    case 'OBJECT':
    case 'SCALAR':
    case 'INPUT_OBJECT':
    case 'INTERFACE': {
      return type
    }
  }
}
