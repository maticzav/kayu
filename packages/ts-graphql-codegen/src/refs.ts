/**
 * This file contains helper methods that we use for code generation
 * and serve a single purpose - that's why I extracted them here.
 */
import { IntrospectionTypeRef } from 'graphql'

import { IntrospectionInvertedTypeRef, invert } from './ast'

/**
 * Wraps the given type using a given reference.
 *
 * @param type - The chunk that we are wrapping.
 * @param ref - The reference that we use as a wrapping base.
 * @param optionals - Tells whether `null` values may be undefined.
 */
export function wrap(
  type: string,
  ref: IntrospectionTypeRef,
  optionals: boolean = false,
): string {
  return wrapInverted(type, invert(ref), optionals)
}

/**
 * This is an internal helper function that lets us more easily define the return type
 * of the refered one.
 */
function wrapInverted(
  type: string,
  ref: IntrospectionInvertedTypeRef,
  optionals: boolean,
): string {
  switch (ref.kind) {
    /* Nullable */
    case 'NULLABLE': {
      const wrapped = wrapInverted(type, ref.ofType, optionals)

      if (optionals) return `(${wrapped} | null | undefined)`
      return `(${wrapped} | null)`
    }
    /* List */
    case 'LIST': {
      return `(${wrapInverted(type, ref.ofType, optionals)}[])`
    }
    /* Named */
    case 'ENUM':
    case 'UNION':
    case 'OBJECT':
    case 'SCALAR':
    case 'INPUT_OBJECT':
    case 'INTERFACE': {
      return type
    }
  }
}
