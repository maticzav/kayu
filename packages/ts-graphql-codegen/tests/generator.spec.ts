import * as fs from 'fs'
import { IntrospectionSchema } from 'graphql'
import * as path from 'path'
import { promisify } from 'util'

import { GQLGenerator, loadSchemaFromPath, loadSchemaFromURL } from '../src'
import { defined } from '../src/utils'

const writefile = promisify(fs.writeFile)
const prettier = require('../../../prettier.config')

/* Configuration */

const ENDPOINT = 'http://localhost:4000/graphql'

const FIXTURES = path.resolve(__dirname, './__fixtures__')
const SCHEMA_PATH = path.resolve(FIXTURES, './schema.json')

const CODECS_PATH = `./${path.resolve(FIXTURES, './codecs')}`
const CORE_PATH = path.resolve(__dirname, '../../ts-graphql/src/')

/* Documentation */

describe('generator', () => {
  let schema: IntrospectionSchema

  beforeAll(async () => {
    schema = await loadSchemaFromPath(SCHEMA_PATH)
  })

  test('generates the API with codecs', async () => {
    /**
     * Construct the generator.
     */
    const generator = new GQLGenerator(schema, prettier)

    /**
     * Save the code and perform tests as described above.
     */
    const code = generator.generate({
      core: path.relative(FIXTURES, CORE_PATH),
      codecs: path.relative(FIXTURES, CODECS_PATH),
    })
    await writefile(path.resolve(FIXTURES, './api-codecs.ts'), code)

    expect(code).toMatchSnapshot()
  })

  test('generates the API without codecs', async () => {
    /**
     * Construct the generator.
     */
    const generator = new GQLGenerator(schema, prettier)

    /**
     * Save the code and perform tests as described above.
     */
    const code = generator.generate({
      core: path.relative(FIXTURES, CORE_PATH),
    })
    await writefile(path.resolve(FIXTURES, './api-nocodecs.ts'), code)

    expect(code).toMatchSnapshot()
  })
})
