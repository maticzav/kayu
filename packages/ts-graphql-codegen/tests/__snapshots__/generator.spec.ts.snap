// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator generates the API 1`] = `
"import {
  composite,
  leaf,
  fragment,
  SelectionSet,
  Fields,
  selection,
  Argument,
  hash,
  arg,
} from 'ts-graphql/src/__generator'

/* Scalars */
export type Scalar = {
  ID: string
  String: string
  Float: number
  Int: number
  Bool: boolean
}

/* Objects */
type DroidObject = {
  id: Scalar['ID'] | null
  name: Scalar['String'] | null
  primaryFunction: Scalar['String'] | null
  appearsIn: Array<Enum['Episode'] | null> | null
}

type HumanObject = {
  id: Scalar['ID'] | null
  name: Scalar['String'] | null
  homePlanet: Scalar['String'] | null
  appearsIn: Array<Enum['Episode'] | null> | null
  infoURL: Scalar['String'] | null
}

type QueryObject = {
  human: Object['Human'] | null
  droid: Object['Droid'] | null
  character: Union['CharacterUnion'] | null
  luke: Object['Human'] | null
  humans: Array<Object['Human'] | null> | null
  droids: Array<Object['Droid'] | null> | null
  characters: Array<Interface['Character'] | null> | null
  greeting: Scalar['String'] | null
  whoami: Scalar['String'] | null
}

export type Object = {
  Droid: DroidObject
  Human: HumanObject
  Query: QueryObject
}

/* Enums */
enum EpisodeEnum {
  NEWHOPE = 'NEWHOPE',
  EMPIRE = 'EMPIRE',
  JEDI = 'JEDI',
}

enum LanguageEnum {
  EN = 'EN',
  SL = 'SL',
}

export type Enum = {
  Episode: EpisodeEnum
  Language: LanguageEnum
}

/* Input Objects */
type GreetingInputObject = {
  language: Enum['Language'] | null | undefined
  name: Scalar['String'] | null | undefined
}

type GreetingOptionsInputObject = {
  prefix: Scalar['String'] | null | undefined
}

export type InputObject = {
  Greeting: GreetingInputObject
  GreetingOptions: GreetingOptionsInputObject
}

/* Field Types */
type FieldsTypes = {
  Droid: {
    id: () => Object['Droid'] | null
    name: () => Object['Droid'] | null
    primaryFunction: () => Object['Droid'] | null
    appearsIn: () => Array<Object['Droid'] | null> | null
  }
  Human: {
    id: () => Object['Human'] | null
    name: () => Object['Human'] | null
    homePlanet: () => Object['Human'] | null
    appearsIn: () => Array<Object['Human'] | null> | null
    infoUrl: () => Object['Human'] | null
  }
  Query: {
    human: (params: {
      id: Scalar['ID']
    }) => <T>(
      selection: SelectionSet<Object['Human'] | null | undefined, T>,
    ) => T
    droid: (params: {
      id: Scalar['ID']
    }) => <T>(
      selection: SelectionSet<Object['Droid'] | null | undefined, T>,
    ) => T
    character: (params: {
      id: Scalar['ID']
    }) => <T>(
      selection: SelectionSet<Union['CharacterUnion'] | null | undefined, T>,
    ) => T
    luke: <T>(
      selection: SelectionSet<Object['Human'] | null | undefined, T>,
    ) => T
    humans: <T>(
      selection: SelectionSet<
        Array<Object['Human'] | null | undefined> | null | undefined,
        T
      >,
    ) => T
    droids: <T>(
      selection: SelectionSet<
        Array<Object['Droid'] | null | undefined> | null | undefined,
        T
      >,
    ) => T
    characters: <T>(
      selection: SelectionSet<
        Array<Interface['Character'] | null | undefined> | null | undefined,
        T
      >,
    ) => T
    greeting: (params: {
      input: InputObject['Greeting'] | null
    }) => Object['Query'] | null
    whoami: () => Object['Query'] | null
  }
}

/* Selections */
export const objects = {
  droid: <T>(
    selector: (fields: FieldsTypes['Droid']) => T,
  ): SelectionSet<Object['Droid'], T> => {
    const decoder = (fields: Fields<Object['Droid']>): T => {
      const types: FieldsTypes['Droid'] = {
        /* id */
        id: () => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          fields.select(leaf('id', args))
          let mock = null

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              return data.response.get('id')(argsHash)
          }
        },
        /* name */
        name: () => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          fields.select(leaf('name', args))
          let mock = null

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              return data.response.get('name')(argsHash)
          }
        },
        /* primaryFunction */
        primaryFunction: () => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          fields.select(leaf('primaryFunction', args))
          let mock = null

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              return data.response.get('primaryFunction')(argsHash)
          }
        },
        /* appearsIn */
        appearsIn: () => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          fields.select(leaf('appearsIn', args))
          let mock = null

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              return data.response.get('appearsIn')(argsHash)
          }
        },
      }
      return selector(types)
    }
    return selection(decoder)
  },
  human: <T>(
    selector: (fields: FieldsTypes['Human']) => T,
  ): SelectionSet<Object['Human'], T> => {
    const decoder = (fields: Fields<Object['Human']>): T => {
      const types: FieldsTypes['Human'] = {
        /* id */
        id: () => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          fields.select(leaf('id', args))
          let mock = null

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              return data.response.get('id')(argsHash)
          }
        },
        /* name */
        name: () => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          fields.select(leaf('name', args))
          let mock = null

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              return data.response.get('name')(argsHash)
          }
        },
        /* homePlanet */
        homePlanet: () => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          fields.select(leaf('homePlanet', args))
          let mock = null

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              return data.response.get('homePlanet')(argsHash)
          }
        },
        /* appearsIn */
        appearsIn: () => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          fields.select(leaf('appearsIn', args))
          let mock = null

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              return data.response.get('appearsIn')(argsHash)
          }
        },
        /* infoUrl */
        infoUrl: () => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          fields.select(leaf('infoURL', args))
          let mock = null

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              return data.response.get('infoURL')(argsHash)
          }
        },
      }
      return selector(types)
    }
    return selection(decoder)
  },
  query: <T>(
    selector: (fields: FieldsTypes['Query']) => T,
  ): SelectionSet<Object['Query'], T> => {
    const decoder = (fields: Fields<Object['Query']>): T => {
      const types: FieldsTypes['Query'] = {
        /* human */
        human: (params) => (selection) => {
          /* Arguments */
          const args: Argument[] = [arg('id', 'ID!', params.id)]
          const argsHash = hash(args)

          /* Selection */
          let subfields = new Fields<Object['Human'] | null>()
          let mock = null
          fields.select(composite('human', subfields.selection, args))

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              let datasubfields = new Fields<Object['Human'] | null>(
                data.response.get('human')(argsHash),
              )
              return selection.decoder(datasubfields)
          }
        },
        /* droid */
        droid: (params) => (selection) => {
          /* Arguments */
          const args: Argument[] = [arg('id', 'ID!', params.id)]
          const argsHash = hash(args)

          /* Selection */
          let subfields = new Fields<Object['Droid'] | null>()
          let mock = null
          fields.select(composite('droid', subfields.selection, args))

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              let datasubfields = new Fields<Object['Droid'] | null>(
                data.response.get('droid')(argsHash),
              )
              return selection.decoder(datasubfields)
          }
        },
        /* character */
        character: (params) => (selection) => {
          /* Arguments */
          const args: Argument[] = [arg('id', 'ID!', params.id)]
          const argsHash = hash(args)

          /* Selection */
          let subfields = new Fields<Union['CharacterUnion'] | null>()
          let mock = null
          fields.select(composite('character', subfields.selection, args))

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              let datasubfields = new Fields<Union['CharacterUnion'] | null>(
                data.response.get('character')(argsHash),
              )
              return selection.decoder(datasubfields)
          }
        },
        /* luke */
        luke: (selection) => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          let subfields = new Fields<Object['Human'] | null>()
          let mock = null
          fields.select(composite('luke', subfields.selection, args))

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              let datasubfields = new Fields<Object['Human'] | null>(
                data.response.get('luke')(argsHash),
              )
              return selection.decoder(datasubfields)
          }
        },
        /* humans */
        humans: (selection) => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          let subfields = new Fields<Array<Object['Human'] | null> | null>()
          let mock = null
          fields.select(composite('humans', subfields.selection, args))

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              let datasubfields = new Fields<Array<
                Object['Human'] | null
              > | null>(data.response.get('humans')(argsHash))
              return selection.decoder(datasubfields)
          }
        },
        /* droids */
        droids: (selection) => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          let subfields = new Fields<Array<Object['Droid'] | null> | null>()
          let mock = null
          fields.select(composite('droids', subfields.selection, args))

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              let datasubfields = new Fields<Array<
                Object['Droid'] | null
              > | null>(data.response.get('droids')(argsHash))
              return selection.decoder(datasubfields)
          }
        },
        /* characters */
        characters: (selection) => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          let subfields = new Fields<Array<
            Interface['Character'] | null
          > | null>()
          let mock = null
          fields.select(composite('characters', subfields.selection, args))

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              let datasubfields = new Fields<Array<
                Interface['Character'] | null
              > | null>(data.response.get('characters')(argsHash))
              return selection.decoder(datasubfields)
          }
        },
        /* greeting */
        greeting: (params) => {
          /* Arguments */
          const args: Argument[] = [arg('input', 'Greeting', params.input)]
          const argsHash = hash(args)

          /* Selection */
          fields.select(leaf('greeting', args))
          let mock = null

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              return data.response.get('greeting')(argsHash)
          }
        },
        /* whoami */
        whoami: () => {
          /* Arguments */
          const args: Argument[] = []
          const argsHash = hash(args)

          /* Selection */
          fields.select(leaf('whoami', args))
          let mock = null

          /* Decoder */
          const data = fields.data
          switch (data.type) {
            case 'fetching':
              return mock
            case 'fetched':
              return data.response.get('whoami')(argsHash)
          }
        },
      }
      return selector(types)
    }
    return selection(decoder)
  },
}
"
`;
